File: yeet-synth/synth/chaos.py
"""
Fault injection and chaos helpers.

These are OPTIONAL and ADDITIVE — they can be mixed into any traffic profile
by enabling chaos_enabled in the profile config.

All chaos scenarios are designed to be safe in staging:
  - They don't delete data
  - They don't bypass security controls that should remain active
  - They produce observable telemetry to verify alert pipelines

WARNING: Do NOT run chaos mode in production without explicit authorisation
and an active incident response channel open.
"""

from __future__ import annotations

import asyncio
import logging
import uuid
from dataclasses import dataclass

import synth.endpoints as ep

from .client import SynthClient
from .payloads import malformed_bet_payload, malformed_login_payload, stale_token
from .token_manager import TokenPool

logger = logging.getLogger(__name__)


@dataclass
class ChaosScenarioResult:
    scenario: str
    status_code: int
    expected_status: int
    passed: bool
    latency_ms: float
    note: str = ""


class ChaosInjector:
    """
    Runs a battery of chaos scenarios against the API.
    Returns a list of ChaosScenarioResults for evaluation.
    """

    def __init__(self, client: SynthClient, pool: TokenPool) -> None:
        self._client = client
        self._pool = pool

    async def run_all(self) -> list[ChaosScenarioResult]:
        results: list[ChaosScenarioResult] = []
        creds = self._pool.get_random()

        if creds:
            results += await self._stale_token_requests(creds.user_id)
            results += await self._malformed_payloads(creds.access_token)
            results += await self._duplicate_idempotency_replay(creds)
            results += await self._rate_limit_burst()
            results += await self._missing_idempotency_key(creds.access_token)
        else:
            logger.warning(
                "Chaos: no token available, skipping auth-required scenarios"
            )

        results += await self._unauthenticated_protected_endpoints()
        return results

    # ── 1. Stale / invalid token ──────────────────────────────────────────────

    async def _stale_token_requests(self, user_id: str) -> list[ChaosScenarioResult]:
        """All requests with an expired token must 401."""
        token = stale_token()
        r = await ep.users_profile(self._client, token, user_id)
        return [
            ChaosScenarioResult(
                scenario="stale_token",
                status_code=r.status_code,
                expected_status=401,
                passed=r.status_code == 401,
                latency_ms=r.latency_ms,
                note="Expired JWT must be rejected with 401",
            )
        ]

    # ── 2. Malformed payloads ─────────────────────────────────────────────────

    async def _malformed_payloads(self, token: str) -> list[ChaosScenarioResult]:
        results = []

        # Malformed bet
        r = await self._client.request(
            "POST",
            "/api/v1/bets/place",
            token=token,
            idempotency_key=str(uuid.uuid4()),
            json=malformed_bet_payload(),
            endpoint_name="POST /api/v1/bets/place [malformed]",
        )
        results.append(
            ChaosScenarioResult(
                scenario="malformed_bet_payload",
                status_code=r.status_code,
                expected_status=400,
                passed=r.status_code == 400,
                latency_ms=r.latency_ms,
                note="Invalid bet payload must return 400 VALIDATION_ERROR",
            )
        )

        # Malformed login
        r = await self._client.request(
            "POST",
            "/api/v1/auth/token",
            json=malformed_login_payload(),
            endpoint_name="POST /api/v1/auth/token [malformed]",
        )
        results.append(
            ChaosScenarioResult(
                scenario="malformed_login_payload",
                status_code=r.status_code,
                expected_status=400,
                passed=r.status_code == 400,
                latency_ms=r.latency_ms,
                note="Invalid login payload must return 400",
            )
        )

        # Oversized body (send 1001 bytes of garbage)
        garbage = {"data": "x" * 1001}
        r = await self._client.request(
            "POST",
            "/api/v1/auth/token",
            json=garbage,
            endpoint_name="POST /api/v1/auth/token [oversized]",
        )
        results.append(
            ChaosScenarioResult(
                scenario="oversized_body",
                status_code=r.status_code,
                expected_status=400,
                passed=r.status_code in (400, 413, 422),
                latency_ms=r.latency_ms,
                note="Oversized malformed body should be rejected",
            )
        )

        return results

    # ── 3. Idempotency replay ─────────────────────────────────────────────────

    async def _duplicate_idempotency_replay(self, creds) -> list[ChaosScenarioResult]:
        """Send the same request twice with the same idempotency key.
        Second must replay."""
        key = str(uuid.uuid4())
        r1 = await ep.wallet_deposit(
            self._client,
            creds.access_token,
            creds.user_id,
            amount="5.00",
            idempotency_key=key,
        )
        await asyncio.sleep(0.1)
        r2 = await ep.wallet_deposit(
            self._client,
            creds.access_token,
            creds.user_id,
            amount="5.00",
            idempotency_key=key,
        )

        return [
            ChaosScenarioResult(
                scenario="idempotency_replay",
                status_code=r2.status_code,
                expected_status=r1.status_code,
                passed=r2.idempotency_replay or r2.status_code == r1.status_code,
                latency_ms=r2.latency_ms,
                note=(
                    f"Replay detected={r2.idempotency_replay}; "
                    f"first={r1.status_code} second={r2.status_code}"
                ),
            )
        ]

    # ── 4. Rate limit burst ───────────────────────────────────────────────────

    async def _rate_limit_burst(self) -> list[ChaosScenarioResult]:
        """Fire > RATE_LIMIT_MAX requests in a short window to trigger 429."""
        tasks = [ep.health_live(self._client) for _ in range(120)]
        records = await asyncio.gather(*tasks)
        rate_limited = [r for r in records if r.status_code == 429]
        return [
            ChaosScenarioResult(
                scenario="rate_limit_trigger",
                status_code=429 if rate_limited else 200,
                expected_status=429,
                passed=len(rate_limited) > 0,
                latency_ms=0.0,
                note=f"{len(rate_limited)}/120 requests returned 429",
            )
        ]

    # ── 5. Missing idempotency key on guarded endpoint ────────────────────────

    async def _missing_idempotency_key(self, token: str) -> list[ChaosScenarioResult]:
        """Omit the Idempotency-Key header on an endpoint that requires it."""
        r = await self._client.request(
            "POST",
            "/api/v1/bets/place",
            token=token,
            # No idempotency_key= argument
            json={
                "game_id": "game_crash_v1",
                "amount": "1.00",
                "currency": "USD",
                "bet_type": "auto_cashout",
            },
            endpoint_name="POST /api/v1/bets/place [no-idem-key]",
        )
        return [
            ChaosScenarioResult(
                scenario="missing_idempotency_key",
                status_code=r.status_code,
                expected_status=400,
                passed=r.status_code == 400,
                latency_ms=r.latency_ms,
                note="Missing Idempotency-Key must return 400 MISSING_IDEMPOTENCY_KEY",
            )
        ]

    # ── 6. Unauthenticated access to protected endpoints ─────────────────────

    async def _unauthenticated_protected_endpoints(self) -> list[ChaosScenarioResult]:
        results = []
        protected = [
            ("/api/v1/bets/history", "GET"),
            ("/api/v1/wallet/00000000-0000-0000-0000-000000000000/balance", "GET"),
            ("/api/v1/config/flags", "GET"),
            ("/_internal/status", "GET"),
        ]
        for path, method in protected:
            r = await self._client.request(
                method,
                path,
                endpoint_name=f"{method} {path} [no-auth]",
                # No token
            )
            results.append(
                ChaosScenarioResult(
                    scenario=f"unauth_{path.split('/')[-1]}",
                    status_code=r.status_code,
                    expected_status=401,
                    passed=r.status_code in (401, 403),
                    latency_ms=r.latency_ms,
                    note=f"{method} {path} without auth must 401/403",
                )
            )
        return results

File: y_eet-synth/internal/chaos/chaos.go
// Package chaos implements fault injection scenarios for staging validation.
// WARNING: Run only in staging or controlled environments.
package chaos

import (
	"context"
	"fmt"
	"math/rand"

	"y_eet-synth/internal/client"
	"y_eet-synth/internal/evaluator"
	"y_eet-synth/internal/token"
)

// Injector runs chaos / fault-path validation scenarios.
type Injector struct {
	cl   *client.Client
	pool *token.Pool
}

// New creates an Injector.
func New(cl *client.Client, pool *token.Pool) *Injector {
	return &Injector{cl: cl, pool: pool}
}

// RunAll executes all chaos scenarios and returns their results.
func (i *Injector) RunAll(ctx context.Context) []evaluator.ChaosResult {
	return []evaluator.ChaosResult{
		i.staleToken(ctx),
		i.malformedPayload(ctx),
		i.duplicateReplay(ctx),
		i.missingIdempotencyKey(ctx),
		i.oversizedPayload(ctx),
	}
}

// staleToken sends a request with a deliberately invalid token.
// The API must respond with 401.
func (i *Injector) staleToken(ctx context.Context) evaluator.ChaosResult {
	r := i.cl.Do(ctx, client.RequestOptions{
		Method:       "GET",
		Path:         "/api/v1/users/chaos-test/profile",
		Token:        "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.INVALID.STALE",
		EndpointName: "GET /api/v1/users/:id/profile",
	})
	passed := r.StatusCode == 401 || r.StatusCode == 403
	return evaluator.ChaosResult{
		Scenario:       "stale_token",
		Passed:         passed,
		ExpectedStatus: 401,
		StatusCode:     r.StatusCode,
		Note: func() string {
			if passed {
				return "stale token correctly rejected"
			}
			return fmt.Sprintf("expected 401/403, got %d — auth middleware may not be enforcing JWT expiry", r.StatusCode)
		}(),
	}
}

// malformedPayload sends an invalid JSON body; API must respond with 400/422.
func (i *Injector) malformedPayload(ctx context.Context) evaluator.ChaosResult {
	creds := i.pool.GetRandom()
	if creds == nil {
		return evaluator.ChaosResult{
			Scenario: "malformed_payload",
			Passed:   true,
			Note:     "skipped: no token available",
		}
	}
	// Send a string where an object is expected
	r := i.cl.Do(ctx, client.RequestOptions{
		Method:       "POST",
		Path:         "/api/v1/bets/place",
		Token:        creds.AccessToken,
		Body:         "this is not json at all ><",
		EndpointName: "POST /api/v1/bets/place",
	})
	passed := r.StatusCode == 400 || r.StatusCode == 422 || r.StatusCode == 415
	return evaluator.ChaosResult{
		Scenario:       "malformed_payload",
		Passed:         passed,
		ExpectedStatus: 400,
		StatusCode:     r.StatusCode,
		Note: func() string {
			if passed {
				return "malformed payload correctly rejected"
			}
			return fmt.Sprintf("expected 400/422, got %d — input validation may not be enforced", r.StatusCode)
		}(),
	}
}

// duplicateReplay sends the same idempotency key twice.
// The second request must return the same 2xx (idempotency) or 409.
func (i *Injector) duplicateReplay(ctx context.Context) evaluator.ChaosResult {
	creds := i.pool.GetRandom()
	if creds == nil {
		return evaluator.ChaosResult{
			Scenario: "duplicate_replay",
			Passed:   true,
			Note:     "skipped: no token available",
		}
	}

	idemKey := fmt.Sprintf("chaos-replay-%08x", rand.Int63()) //nolint:gosec
	body := map[string]interface{}{
		"amount":           "10.00",
		"currency":         "USD",
		"payment_method":   "card",
		"client_reference": idemKey,
	}

	uid := creds.UserID
	opts := client.RequestOptions{
		Method:         "POST",
		Path:           "/api/v1/wallet/" + uid + "/deposit",
		Token:          creds.AccessToken,
		IdempotencyKey: idemKey,
		Body:           body,
		EndpointName:   "POST /api/v1/wallet/:id/deposit",
	}

	r1 := i.cl.Do(ctx, opts)
	r2 := i.cl.Do(ctx, opts) // exact replay

	// Second request should be idempotent (same 2xx) or 409
	passed := (r2.StatusCode >= 200 && r2.StatusCode < 300) ||
		r2.StatusCode == 409 || r2.IdempotencyReplay
	return evaluator.ChaosResult{
		Scenario:       "duplicate_replay",
		Passed:         passed,
		ExpectedStatus: r1.StatusCode,
		StatusCode:     r2.StatusCode,
		Note: func() string {
			if passed {
				return fmt.Sprintf("idempotency replay handled correctly (r1=%d, r2=%d)", r1.StatusCode, r2.StatusCode)
			}
			return fmt.Sprintf("idempotency not enforced: r1=%d, r2=%d", r1.StatusCode, r2.StatusCode)
		}(),
	}
}

// missingIdempotencyKey omits the key on an endpoint that requires it.
func (i *Injector) missingIdempotencyKey(ctx context.Context) evaluator.ChaosResult {
	creds := i.pool.GetRandom()
	if creds == nil {
		return evaluator.ChaosResult{
			Scenario: "missing_idempotency_key",
			Passed:   true,
			Note:     "skipped: no token available",
		}
	}

	r := i.cl.Do(ctx, client.RequestOptions{
		Method: "POST",
		Path:   "/api/v1/wallet/" + creds.UserID + "/deposit",
		Token:  creds.AccessToken,
		// Deliberately no IdempotencyKey
		Body: map[string]interface{}{
			"amount": "5.00", "currency": "USD", "payment_method": "card",
		},
		EndpointName: "POST /api/v1/wallet/:id/deposit",
	})

	// API may require the key (400) or tolerate its absence (2xx).
	// Either is valid; we just verify no 5xx.
	passed := r.StatusCode < 500
	return evaluator.ChaosResult{
		Scenario:       "missing_idempotency_key",
		Passed:         passed,
		ExpectedStatus: 400,
		StatusCode:     r.StatusCode,
		Note: func() string {
			if r.StatusCode == 400 {
				return "missing idempotency key correctly rejected with 400"
			}
			if r.StatusCode >= 200 && r.StatusCode < 300 {
				return "idempotency key optional on this endpoint (accepted without it)"
			}
			if !passed {
				return fmt.Sprintf("unexpected %d — server error on missing idempotency key", r.StatusCode)
			}
			return fmt.Sprintf("status %d", r.StatusCode)
		}(),
	}
}

// oversizedPayload sends a very large body to probe request size limits.
func (i *Injector) oversizedPayload(ctx context.Context) evaluator.ChaosResult {
	creds := i.pool.GetRandom()
	if creds == nil {
		return evaluator.ChaosResult{
			Scenario: "oversized_payload",
			Passed:   true,
			Note:     "skipped: no token available",
		}
	}

	// 1 MB of noise in a JSON field
	noise := make([]byte, 1024*1024)
	for j := range noise {
		noise[j] = 'A'
	}
	r := i.cl.Do(ctx, client.RequestOptions{
		Method:       "POST",
		Path:         "/api/v1/bets/place",
		Token:        creds.AccessToken,
		Body:         map[string]interface{}{"junk": string(noise)},
		EndpointName: "POST /api/v1/bets/place",
	})

	passed := r.StatusCode == 413 || r.StatusCode == 400 || r.StatusCode == 422
	return evaluator.ChaosResult{
		Scenario:       "oversized_payload",
		Passed:         passed,
		ExpectedStatus: 413,
		StatusCode:     r.StatusCode,
		Note: func() string {
			if passed {
				return fmt.Sprintf("oversized payload rejected with %d", r.StatusCode)
			}
			return fmt.Sprintf("server accepted 1MB payload (%d) — check Nginx/Istio body size limits", r.StatusCode)
		}(),
	}
}
